pulp/apis/cuopt_api.py

```diff
@@ -0,0 +1,337 @@
+    sparse,
+)
+
+# COPT string convention
+if sys.version_info >= (3, 0):
+    coptstr = lambda x: bytes(x, "utf-8")
+else:
+    coptstr = lambda x: x
+
+byref = ctypes.byref
+
+# Constraint Sense Converter
+sense_conv = {LpConstraintLE: "L",
+              LpConstraintGE: "G",
+              LpConstraintEQ: "E",
+             }
+
+def process_constraint(args):
+    local_data = []
+    local_vars = []
+    local_sense = []
+    local_rhs = []
+    for constraints in args:
+        name, constraint = constraints
+        local_data.extend(list(constraint.values()))
+        local_vars.extend(list(constraint.keys()))
+        local_sense.append(sense_conv[constraint.sense])
+        local_rhs.append(-constraint.constant)
+    return local_data, local_vars, local_sense, local_rhs
+
+
+def process_variables(var):
+    #obj_coeff = lp.objective.get(var, 0.0))
+    lowBound = var.lowBound
+    if lowBound is None:
+        lowBound = -np.inf
+    upBound = var.upBound
+    if upBound is None:
+        upBound = np.inf
+    varType = "C"
+    if var.cat == LpInteger and self.mip:
+        varType = "I"
+    if var.cat == LpBinary and self.mip:
+        varType = "I"
+        lowBound = 0
+        upBound  = 1
+    return lowBound, upBound, varType, var.name
+
+
+class CUOPT(LpSolver):
+    """
+    The CUOPT Optimizer via its python interface
+    """
+
+    name = "CUOPT"
+
+    try:
+        global cuopt
+        import cuopt  # type: ignore[import-not-found]
+    except:
+
+        def available(self):
+            """True if the solver is available"""
+            return False
+
+        def actualSolve(self, lp, callback=None):
+            """Solve a well formulated lp problem"""
+            raise PulpSolverError("CUOPT: Not available")
+
+    else:
+
+        def __init__(
+            self,
+            mip=True,
+            msg=True,
+            timeLimit=None,
+            gapRel=None,
+            warmStart=False,
+            logPath=None,
+            **solverParams,
+        ):
+            """
+            :param bool mip: if False, assume LP even if integer variables
+            :param bool msg: if False, no log is shown
+            :param float timeLimit: maximum time for solver (in seconds)
+            :param float gapRel: relative gap tolerance for the solver to stop (in fraction)
+            :param bool warmStart: if True, the solver will use the current value of variables as a start
+            :param str logPath: path to the log file
+            """
+
+            LpSolver.__init__(
+                self,
+                mip=mip,
+                msg=msg,
+                timeLimit=timeLimit,
+                gapRel=gapRel,
+                logPath=logPath,
+                warmStart=warmStart,
+            )
+
+            from cuopt.linear_programming import data_model, solver_settings, solver
+            self.model = data_model.DataModel()
+            self.settings = solver_settings.SolverSettings()
```

Comment:
why is it needed to define the settings during the initialization? I think all the settings management can be done inside the `callSolver` method without the need to creating and maintaining a `settings` attribute.

Reply:
Moved solver_settings to CallSolver. Refer #839